staging/src/k8s.io/apiserver/pkg/registry/rest/validate.go

```diff
@@ -71,30 +111,56 @@ func ValidateDeclaratively(ctx context.Context, options sets.Set[string], scheme
 // before validation occurs. The scheme MUST have the declarative validation
 // registered for the requested resource/subresource.
 //
-// option should contain any validation options that the declarative validation
-// tags expect.
-//
 // Returns a field.ErrorList containing any validation errors. An internal error
 // is included if requestInfo is missing from the context or if version
 // conversion fails.
-func ValidateUpdateDeclaratively(ctx context.Context, options sets.Set[string], scheme *runtime.Scheme, obj, oldObj runtime.Object) field.ErrorList {
-	if requestInfo, found := genericapirequest.RequestInfoFrom(ctx); found {
-		groupVersion := schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}
-		versionedObj, err := scheme.ConvertToVersion(obj, groupVersion)
+func ValidateUpdateDeclaratively(ctx context.Context, scheme *runtime.Scheme, obj, oldObj runtime.Object, configOpts ...ValidationConfig) field.ErrorList {
+	cfg := &validationConfigOption{opType: operation.Update}
+	for _, o := range configOpts {
+		o(cfg)
+	}
+	return panicSafeValidateFunc(validateDeclaratively, cfg.takeover)(ctx, scheme, obj, oldObj, cfg)
+}
+
+func validateDeclaratively(ctx context.Context, scheme *runtime.Scheme, obj, oldObj runtime.Object, o *validationConfigOption) field.ErrorList {
+	// Find versionedGroupVersion, which identifies the API version to use for declarative validation.
+	versionedGroupVersion, subresources, err := requestInfo(ctx, o.subresourceGVKMapper)
+	if err != nil {
+		return field.ErrorList{field.InternalError(nil, err)}
+	}
+	versionedObj, err := scheme.ConvertToVersion(obj, versionedGroupVersion)
+	if err != nil {
+		return field.ErrorList{field.InternalError(nil, fmt.Errorf("unexpected error converting to versioned type: %w", err))}
+	}
+	var versionedOldObj runtime.Object
+
+	switch o.opType {
+	case operation.Create:
+		return scheme.Validate(ctx, o.options, versionedObj, subresources...)
+	case operation.Update:
+		versionedOldObj, err = scheme.ConvertToVersion(oldObj, versionedGroupVersion)
 		if err != nil {
 			return field.ErrorList{field.InternalError(nil, fmt.Errorf("unexpected error converting to versioned type: %w", err))}
 		}
-		versionedOldObj, err := scheme.ConvertToVersion(oldObj, groupVersion)
-		if err != nil {
-			return field.ErrorList{field.InternalError(nil, fmt.Errorf("unexpected error converting to versioned type: %w", err))}
+		return scheme.ValidateUpdate(ctx, o.options, versionedObj, versionedOldObj, subresources...)
+	default:
+		return field.ErrorList{field.InternalError(nil, fmt.Errorf("unknown operation type: %v", o.opType))}
+	}
+}
+
+func requestInfo(ctx context.Context, subresourceMapper GroupVersionKindProvider) (schema.GroupVersion, []string, error) {
+	if requestInfo, found := genericapirequest.RequestInfoFrom(ctx); found {
```

Comment:
nit: inverting to `!found` allows the early return and then all the logic outside the nesting.